pageserver/src/page_service.rs

```diff
@@ -370,6 +370,12 @@ impl PageServerHandler {
     ) -> anyhow::Result<()> {
         let _enter = info_span!("pagestream", timeline = %timelineid, tenant = %tenantid).entered();
 
+        thread_mgr::associate_with(Some(tenantid), Some(timelineid));
+        // reset on handler exit
+        scopeguard::defer! {
+            thread_mgr::associate_with(None, None);
+        }
+
```

Comment:
Hmm, I don't think it's possible to stop the pagestream and switch to another tenant, within the same connection. So this scopeguard isn't really needed, you could let the thread be associated with the tenant and timeline until it exits.

Could structure the code to make that more clear, though. Currently, `process_query` calls `handle_pagerequests` the same way as the other `handle_*` functions, even though `handle_pagerequests` will never return normally, without an error nor shutdown. (If it did, it would leave the libpq protocol in a funny state, as it never sends the CopyDone message)

Reply:
Huh, it indeed looks a bit strange. [read_message can return None](https://github.com/neondatabase/neon/blob/main/libs/utils/src/pq_proto.rs#L161) when EOF is reached during message size parsing. Then handler will exit with Ok, and backend code will [try to send ReadyForQuery](https://github.com/neondatabase/neon/blob/main/libs/utils/src/postgres_backend.rs#L443) which should fail because the connection was closed. So yeah, you're right it is not possible, I've removed the defer